File: C++/Convert-Sorted-List-to-Binary-Search-Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	ListNode* getMidPre(ListNode* head) {
		if (head == NULL) return head;
		ListNode *slow = head, *fast = head->next, *prev = NULL;
		while (fast != NULL && fast->next != NULL) {
			prev = slow;
			slow = slow->next;
			fast = fast->next->next;
		}
		if (prev == NULL) return slow;
		return prev;
	}
    TreeNode* sortedListToBST(ListNode* head) {
        if (head == NULL) return NULL;
        if (head->next == NULL) {
        	TreeNode* root = new TreeNode(head->val);
        	return root;
        }
        ListNode* prev = getMidPre(head);
        TreeNode* root = new TreeNode(prev->next->val);
        ListNode* start = prev->next->next;
        prev->next = NULL;
        root->left = sortedListToBST(head);
        root->right = sortedListToBST(start);
        return root;
    }
};

File: Python/Convert-Sorted-List-to-Binary-Search-Tree.py
#!/usr/bin/env python

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a list node
    # @return a tree node
    def sortedListToBST(self, head):
        if head == None:
            return None
        slow = head
        fast = head
        prev = None
        while fast.next != None and fast.next.next != None:
            prev = slow
            fast = fast.next.next
            slow = slow.next
        if head == slow:
            head = None
        if prev != None:
            prev.next = None
        root = TreeNode(slow.val)
        left = self.sortedListToBST(head)
        right = self.sortedListToBST(slow.next)
        root.left = left
        root.right = right
        return root